File: day_03/main.py
import os
from typing import Any, SupportsIndex, overload

def parse(raw: str) -> list[str]:
    return raw.strip().splitlines()


def part1(back_array: list[str]):
    
    total = 0
    for banks in back_array:
        l_largest_index = 0

        for i in range(len(banks)-1):
            
            if banks[i] > banks[l_largest_index]:
                l_largest_index = i

        r_largest_index = l_largest_index + 1
        for i in range(l_largest_index,len(banks)):
            if i == l_largest_index:
                continue
            if banks[i] > banks[r_largest_index]:
                r_largest_index = i
        total += int(banks[l_largest_index]+ banks[r_largest_index])

    print(total)

def part2(bank_array: list[str]):
    def optimize_bank(bank:str, size=12) -> str:
        if size == 0:
            return ""
        max_val = max(bank[:len(bank) - size + 1])
        return max_val + optimize_bank(bank[bank.index(max_val)+1:],size-1)
    total = 0

    for bank in bank_array:
        total += int(optimize_bank(bank))

    print(total)



test = 0
if __name__ == "__main__" and not test:
    dir = os.path.dirname(__file__) + "/data.txt"
    with open(dir, "r") as file:
        data = parse(file.read())
        part2(data)
else:
    data = parse("987654321111111\n"+
    "811111111111119\n"+
    "234234234234278\n"+
    "818181911112111")
    part2(data)

File: day_03/main.cpp
#include <algorithm>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <iterator>
#include <string>
#include <string_view>
#include <vector>
#include <filesystem>

template <typename printable>
void printarray(std::vector<printable> arr) {
    for (auto i: arr) {
        std::cout << i << std::endl;
    }
}

std::vector<std::string> parse_file(std::ifstream& file) {
    std::vector<std::string> output;
    std::string line;
    while (std::getline(file, line)) {
        output.push_back(line);
    }
    return output;
}

std::string optimize_bank(std::string bank, int size=12) {
    if (size == 0) return "";
    if (auto max = std::max_element(bank.begin(), bank.end() - size + 1); max != bank.end() - size + 1)  {
        // std::clog << "max: " << *max << '\n';
     
        auto index = max - bank.begin();
        return std::string{*max} + optimize_bank(
            bank.substr(
                index + 1,
                bank.length() - index + 1
            ), size-1);
    }
    return "";
}

void part2(std::vector<std::string> data){
    unsigned long long total = 0;
    for (auto bank : data) {
        std::string d = optimize_bank(bank);
        // std::cout << bank << " '" << d << "' int:" << std::atoll(d.c_str()) << std::endl;
        total += std::atoll(d.c_str());
    }
    std::cout << total << std::endl;
}
bool test = 0;
int main(int argc, char* argv[]) {
    if (not test) {
        std::filesystem::path exec_path = argv[0];
        auto file = std::ifstream(exec_path.parent_path() / "data.txt");
        if (!file.is_open()) std::cerr << "Error opening the file" << std::endl;
        auto data = parse_file(file);
        part2(data);
    } else {
        std::vector<std::string> data = {"987654321111111","811111111111119","234234234234278","818181911112111"};
        part2(data);
    }

    return 0;
}


